[MLIR][ONNX] Add OnnxToTorch support for Slice Op

wu-s-john opened the pull request.

wu-s-john commented:
Implement https://github.com/nod-ai/SHARK-Turbine/issues/259

@stellaraccident @vivekkhandelwal1 @rsuderman please review

wu-s-john commented:
I don't know why there are testing errors detected tbh.

I looked at failures more carefully and it seems that the `test_transpose_all_permutations_4` produced the wrong code that the test was expecting. However, it actually generated the right outputs. Not sure what's going on. Possibly, a data race is occurring?

stellaraccident commented:
This looks in reasonable shape sans test failures and conflicts. Assigning to either Rob or Vivek to triage / get it landed. (a lot of people are on holiday -- apologies for delays)

rsuderman commented:
I'll pull this down and take a look.

rsuderman commented:
@wu-s-john Managed to fix through the errors in the follow branch:

https://github.com/rsuderman/torch-mlir/tree/slice

Looks like you were throwing an `i64` as the element type for a torch tensor when they enforce signedness. It also appears that `torch-mlir` still uses `-1` for its dynamic shape signifier which triggered the other assertion failure.

As a useful trick, LLVM includes assertions that catch these errors. So `cmake -DLLVM_ENABLE_ASSERTIONS=ON .` often helps track down these failures.

I'll add some extra changes that I would recommend on top of my diff.

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+
+        // Infer Axes
+        int64_t numAxes;
+        if (binder.getNumOperands() >= 4) {
```
Rather than an if-else, set the value to the default then only change if the operand exists.

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+        }
+
+        // Infer Steps
+        if (binder.getNumOperands() >= 5) {
```
Same thing here

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+        // Infer Axes
+        int64_t numAxes;
+        if (binder.getNumOperands() >= 4) {
+          if (binder.tensorOperandAtIndex(axes, 3)) {
```
Bind / get the axes outside of this if and in a check if there are 3 or more operands. It makes it clear that we always get this for different number of operands.

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+                           "the size of starts and ends");
+        }
+
+        auto resultTy =
```
Move this to the top with the rest of the type validation.

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+        if (!resultTy)
+          return failure();
+
+        Location loc = binder.getLoc();
```
Move this to the top with the operand declarations.

wu-s-john commented:
> Looks like you were throwing an i64 as the element type for a torch tensor when they enforce signedness

Ooo thanks for catching this. It seemed like other places in the code were using i64 and it seemed natural to use this. Thanks for the tip for using `cmake -DLLVM_ENABLE_ASSERTIONS=ON`. I will make sure to not have these errors next time.

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+        if (!steps) {
+          defaultStepSize = rewriter.create<Torch::ConstantIntOp>(
+              loc, rewriter.getType<Torch::IntType>(),
+              rewriter.getIntegerAttr(rewriter.getIntegerType(64), 1));
```
Change to 
```
rewriter.getIntegerAttr(rewriter.getIntegerType(64), /*isSigned=*/true));
```

rsuderman commented on lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:
```diff
@@ -870,6 +870,156 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+            endsTorchTy.toBuiltinTensor().dyn_cast<RankedTensorType>();
+
+        if (!operandTy)
+          return failure();
```
You should include an error message

wu-s-john commented:
@rsuderman 

Made corrections to the PR. Addressed most of your comments. Overall, your concern was to simplify the logic for nonbinded arguments and supply them with default ones early one. Implemented these and the code looks a lot cleaner now! I think you would like it!
